### DISCUSSION TITLE
Build error met-10.1.1.20220419 on macOS Monterey 12.1 / apple-clang 13.1.6 / homebrew Python 3.9 / spack

### DISCUSSION THREAD

**Discussion by climbfuji** [dtcenter/METplus#1762] [Installation] [answered]
## Describe the Problem ##
Trying to build met-10.1.1.20220419 on macOS Monterey 12.1 with apple-clang 13.1.6, homebrew Python 3.9 using spack-stack (https://github.com/noaa-emc/spack-stack) in Rosetta2 emulator mode for `x86_64`. Build fails at the very end with a message that I can't make sense of:
```
/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/spack/lib/spack/env/clang/clang++   -L../../../../src/basic/vx_cal -L../../../../src/basic/vx_config -L../../../../src/basic/vx_log -L../../../../src/basic/vx_math -L../../../../src/basic/vx_util -L../../../../src/libcode/vx_afm -L../../../../src/libcode/vx_analysis_util -L../../../../src/libcode/vx_color -L../../../../src/libcode/vx_data2d -L../../../../src/libcode/vx_data2d_factory -L../../../../src/libcode/vx_data2d_grib -L../../../../src/libcode/vx_data2d_grib2 -L../../../../src/libcode/vx_data2d_nccf -L../../../../src/libcode/vx_data2d_nc_met -L../../../../src/libcode/vx_data2d_nc_pinterp -L../../../../src/libcode/vx_data2d_python -L../../../../src/libcode/vx_python3_utils -L../../../../src/libcode/vx_geodesy -L../../../../src/libcode/vx_gis -L../../../../src/libcode/vx_gnomon -L../../../../src/libcode/vx_grid -L../../../../src/libcode/vx_gsl_prob -L../../../../src/libcode/vx_nav -L../../../../src/libcode/vx_nc_obs -L../../../../src/libcode/vx_nc_util -L../../../../src/libcode/vx_pb_util -L../../../../src/libcode/vx_plot_util -L../../../../src/libcode/vx_pointdata_python -L../../../../src/libcode/vx_ps -L../../../../src/libcode/vx_pxm -L../../../../src/libcode/vx_render -L../../../../src/libcode/vx_regrid -L../../../../src/libcode/vx_bool_calc -L../../../../src/libcode/vx_shapedata -L../../../../src/libcode/vx_solar -L../../../../src/libcode/vx_statistics -L../../../../src/libcode/vx_stat_out -L../../../../src/libcode/vx_summary -L../../../../src/libcode/vx_time_series -L../../../../src/libcode/vx_series_data -L../../../../src/libcode/vx_tc_util -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/netcdf-c-4.8.1-vzayhsp/lib -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/netcdf-cxx4-4.3.1-nvhg3ao/lib -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/zlib-1.2.12-e3supdv/lib  -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/g2c-1.6.4-4az7tn6/lib -L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9 -lPython -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/gsl-2.7.1-lj7g63f/lib -L/Users/heinzell/work/spack-stack/spack-stack-update-esmf-from-authoritative/envs/met-test/install/apple-clang/13.1.6/bufr-11.7.0-772urso/lib -o ensemble_stat ensemble_stat-ensemble_stat.o ensemble_stat-ensemble_stat_conf_info.o -lvx_stat_out -lvx_statistics -lvx_shapedata -lvx_gsl_prob -lvx_analysis_util -lvx_data2d_factory -lvx_data2d_nc_met -lvx_data2d_grib -lvx_data2d_grib2 -lg2c -ljasper -lpng -lz -lvx_data2d_nc_pinterp -lvx_data2d_nccf -lvx_data2d_python -lvx_pointdata_python -lvx_python3_utils -L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9 -lPython -lvx_data2d -lvx_nc_obs -lvx_nc_util -lvx_regrid -lvx_grid -lvx_config -lvx_color -lvx_util -lvx_math -lvx_cal -lvx_log -lvx_data2d_python -lvx_pointdata_python -lvx_python3_utils -L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9 -lPython -lm -lnetcdf_c++4 -lnetcdf -lgsl -lgslcblas -lnetcdf -lnetcdf_c++4 -lz
ld: cannot link directly with dylib/framework, your binary is not an allowed client of /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/lib/libPython.tbd for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[5]: *** [ensemble_stat] Error 1
```
I don't see from that lengthy command where the macOS Python framework is coming in, all I see is multiple occurrences of `-L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9 -lPython`. This should be fine, because that's homebrew's Python.

I searched for similar errors on the web, and noticed that there wasn't a lot of information out there, and what I could find didn't really explain why it's trying to use the Apple SDK when Python is actually coming from homebrew.

Just wondering if any of the MET developers have come across this issue by coincidence.

Note. MET builds fine in our GitHub actions CI tests, where we also have macOS builds.

**Comment by JohnHalleyGotway** [comment]
Hi @climbfuji! I see you created an issue in the MET repository about this compilation question. I transferred it over to the METplus repo and converted it to a discussion in the [Installation](https://github.com/dtcenter/METplus/discussions/categories/installation) category. We handle Q&A for all of the METplus components, including MET itself, in the single METplus discussions instance, rather than having separate discussion areas for each component.

I hope you're settling in well at JCSDA and hope there's some potential for the use of METplus there!

@jprestop is the person with the most experience compiling MET on a variety of platforms, so she's a great resource. Although I can confirm that I routinely compile MET during development on my own Mac laptop and don't experience this error. And I haven't heard of this issue from any other users or developers.

Some Googling led me to Stack Overflow and I thought [this comment](https://stackoverflow.com/a/58719834) looked informative. There may be some system setting that's triggering this error.

Checking my own settings though, you'll see that I'm configured in a very similar way:
```
printenv | grep MET_PYTHON_
MET_PYTHON_CC=-I/opt/local/Library/Frameworks/Python.framework/Versions/3.8/include/python3.8
MET_PYTHON_LD=-L/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/config-3.8-darwin -lpython3.8 -ldl -framework CoreFoundation
```
I believe I installed this using MacPorts rather than Homebrew. But one big difference is that I am NOT compiling with clang. Instead, I'm using GNU:
```
CXX=/opt/local/bin/g++
CC=/opt/local/bin/gcc
F77=/opt/local/bin/gfortran
```
And that was also installed via MacPorts (`g++ (MacPorts gcc6 6.5.0_6) 6.5.0`).

@jprestop I wonder if perhaps @prestopUCAR has compiled MET with python embedding support using clang via CLion?

**Reply by jprestop** [reply]
Hi @climbfuji.  I have not successfully compiled on a Mac previously, but I am curious about "-lPython".  In previous compilations, I don't believe I have seen a capitalized "P" in Python for linking, nor the word "python" without a version attached to it.  Note the "-lpython3.8" in @JohnHalleyGotway's MET_PYTHON_LD setting.  

Can you please tell us what you have MET_PYTHON_LD set to?  If it includes "/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9" could you please provide a listing of the contents of that directory?

**Reply by jprestop** [reply]
Hi @climbfuji.  I just wanted to follow up and see if you've had any luck with the compilation?  

If not, could you please tell us what you have MET_PYTHON_LD set to? If it includes "/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9" could you please provide a listing of the contents of that directory?

**Reply by climbfuji** [reply]
Apologies, I've been side tracked working on other issues. I'll try to do this as soon as I get the chance. FYI, I recorded an issue on our spack-stack repo as well - https://github.com/NOAA-EMC/spack-stack/issues/316. There is no more information in it for now, but it's good to have them linked.

**Reply by jprestop** [reply]
Sounds good!  Thank you!

**Reply by jprestop** [reply]
Hi @climbfuji.  I just wanted to check in again to see if you have been able to get back to this work?

**Reply by jprestop** [reply]
Hi @climbfuji. Just following up on this Discussion.  Have you been able to check on your MET_PYTHON_LD setting to check the python library you are linking with and retry the compilation?

**Reply by jprestop** [reply]
> Not yet, sorry. We have a release at the end of this week that has taken up all my time in the last month.

That's ok.  I understand.  I just didn't want to let this Discussion go unanswered for too long, so I thought I'd follow up.  Good luck with the release!

**Comment by prestopUCAR** [comment]
> @jprestop I wonder if perhaps @prestopUCAR has compiled MET with python embedding support using clang via CLion?

Sorry, no.  I did not have my mac or CLion back when I was doing MET development.

**Reply by climbfuji** [reply]
Not yet, sorry. We have a release at the end of this week that has taken up all my time in the last month.

**Comment by climbfuji** [comment, marked as answer]
@jprestop @JohnHalleyGotway I finally got to work on this and found the reason and a solution. It had to do with how `spack` defined the shared Python library. The issue and pull request for the authoritative spack repository are:

https://github.com/spack/spack/issues/33409
https://github.com/spack/spack/pull/33410

The same commits are used in our spack fork:

https://github.com/NOAA-EMC/spack-stack/issues/316
https://github.com/NOAA-EMC/spack/pull/185

**Reply by jprestop** [reply]
That's great news, @climbfuji!  Thank you so much for following up and letting us know.  I'll go ahead and lock this Discussion.  Please feel free to create a new discussion with any other issues that should arise.
